lib/socketclusterclient.rb

```diff
@@ -0,0 +1,263 @@
+require 'websocket-eventmachine-client'
+require 'json'
+require_relative './socketclusterclient/parser'
+require_relative './socketclusterclient/emitter'
+
+#
+# Class Socketcluster provides an interface to create a socket instance
+# use to connect to the socketcluster server
+# @author Maanav Shah <shahmaanav07@gmail.com>
+#
+class Socketcluster
+  include Emitter
+
+  def initialize(url)
+    puts 'initialize got called'
+    @id = ''
+    @cnt = 0
+    @auth_token = nil
+    @url = url
+    @acks = {}
+    @channels = []
+    @enable_reconnection = true
+    @delay = 3
+    @ws = @on_connected = @on_disconnected = @on_connect_error = @on_set_authentication = @on_authentication = nil
+    init
+  end
+
+  def set_basic_listener(on_connected, on_disconnected, on_connect_error)
+    @on_connected = on_connected
+    @on_disconnected = on_disconnected
+    @on_connect_error = on_connect_error
+  end
+
+  def set_authentication_listener(on_set_authentication, on_authentication)
+    @on_set_authentication = on_set_authentication
+    @on_authentication = on_authentication
+  end
+
+  def set_auth_token(token)
+    @auth_token = token.to_s
+  end
+
+  def reset_value
+    @cnt = 0
+  end
+
+  def increment_cnt
+    @cnt += 1
+  end
+
+  def get_subscribed_channels
+    @channels
+  end
+
+  def subscribe_channels
+    @channels.each { |channel| sub(channel) }
+  end
+
+  def Ack(cid)
+    ws = @ws
+
+    def MessageAck(error, data)
+      ack_object = {}
+      ack_object["error"] = error
+      ack_object["data"] = data
+      ack_object["rid"] = cid
+      ws.send(ack_object.as_json)
+    end
+
+    return MessageAck
+  end
+
+  def connect
+    puts 'connect got called'
+
+    EM.epoll
+
+    EM.run do
+      trap('TERM') { stop }
+      trap('INT')  { stop }
+
+      @ws = WebSocket::EventMachine::Client.connect(uri: @url)
+
+      @ws.onopen do
+        puts 'on open got called'
+        reset_value
+        @on_connected.call
```

Comment:
Check if on connected is null or not before calling. Not sure if it throws exception for null values.

Reply:
Also Call to onConnected is supposed to be after sending a handshake. It was causing some problems in python client, So I made changes in the same way.

Reply:
Done.